### PR TITLE
MAPI/CAPI updates

### PR DESCRIPTION
**Pull request by gsrajpal**
Support for all the machine lifecycle actions.

### CODE REVIEW

File: pkg/actuators/client/client.go
Comment (srcarrier) [inline pkg/actuators/client/client.go:50]: Doesn't look as though `resourceControllerService` is being used anywhere in the current code base.
Is it needed?

File: pkg/actuators/client/client.go
Comment (srcarrier) [inline pkg/actuators/client/client.go:61]: Might be worth consolidating IamAuthenticator since the same credentials are passed to more than one service.
```
authenticator := &core.IamAuthenticator{ ApiKey: credentialVal }
vpcService, err := vpcv1.NewVpcV1(&vpcv1.VpcV1Options{
	Authenticator: authenticator,
})
```

File: pkg/actuators/client/client.go
Comment (gsrajpal) [reply pkg/actuators/client/client.go:50]: Yup, not anymore, I left it in there in case we need to use it later, but I guess we can add it then.

I was using it to pull a list of all the resources instances in a resource group.

File: pkg/actuators/client/client.go
Comment (gsrajpal) [reply pkg/actuators/client/client.go:61]: will change it back to how it was :)

File: pkg/actuators/client/client.go
Comment (srcarrier) [inline pkg/actuators/client/client.go:136]: In addition to `err != nil`, we may want to check the response codes from the service as well.
If for some reason the region wasn't found and the service returned a 404, I'm not certain whether err would be nil or not. If err != nil and the return code was 200, we know we're good.

Comment (gsrajpal) [reply pkg/actuators/client/client.go:136]: Good point. I just tried it locally.

In the case of a 404, we should get a `! nil` err with a `region not found`. So we should be good.

File: pkg/actuators/client/client.go
Comment (srcarrier) [inline pkg/actuators/client/client.go:165]: With the `name` filter being designated in the request options, we shouldn't need to check for a name match here.
According to the sdk doc, the `name` parameter:
> "Filters the collection to resources with the exact specified name."

File: pkg/actuators/client/client.go
Comment (srcarrier) [inline pkg/actuators/client/client.go:141]: Since the service endpoint isn't designated in the `NewClient` function, wouldn't this `SetServiceURL` call be needed in all the functions?
I only see it inside `InstanceGetByName` and `InstanceCreate`.
Seems like it'd be nice to designate the service endpoint within `NewClient`, so we don't have to worry about it defaulting to `us-south` in one or more functions. That'd be a potentially tricky problem to debug.

File: pkg/actuators/client/client.go
Comment (srcarrier) [inline pkg/actuators/client/client.go:175]:
```suggestion
// InstanceGetByID retrieves a single instance specified by instanceID
```

File: pkg/actuators/client/client.go
Comment (gsrajpal) [reply pkg/actuators/client/client.go:165]: That was my initial thought as well.

I left the name check in there just to double-check. It should not be needed, but after working with the SDK, I thought there is no harm to have that extra check. 

...should be an easy fix, if we want to do that?

File: pkg/actuators/client/client.go
Comment (srcarrier) [inline pkg/actuators/client/client.go:312]: Wondering if setting the visibility filter here is necessary?
The resource group and name filters should yield a single image result.
Should vpc ever change the visibility options, add support for non-private custom images, or provide a non-private rhcos image for use in the future, this setting could potentially require adjustment later on.
i.e. does it buy us anything?

File: pkg/actuators/client/client.go
Comment (srcarrier) [inline pkg/actuators/client/client.go:331]:
```suggestion
				return *eachImage.ID, nil
```
We probably don't need the `imageID` var

File: pkg/actuators/client/client.go
Comment (gsrajpal) [reply pkg/actuators/client/client.go:141]: For sure -- I just was not sure if for every machine request there is a new client initialization. If that is the case, we can certainly set a few other specifications in the client as well and remove them from other functions.

@JoelSpeed @elmiko Question? 
Does each machine request have a new client initialization or is it one time during controller startup?

File: pkg/actuators/client/client.go
Comment (srcarrier) [inline pkg/actuators/client/client.go:342]:
```suggestion
	options := c.resourceManagerService.NewListResourceGroupsOptions()
	options.SetName(resourceGroupName)
	resourceGroupList, _, err := c.resourceManagerService.ListResourceGroups(options)
```
If we filter by name in the request, we won't need to iterate through all the resource groups for the account or match the names in the `if` statement. Should return a list with just the group we're filtering on.

File: pkg/actuators/client/client.go
Comment (srcarrier) [inline pkg/actuators/client/client.go:354]:
```suggestion
				return *eachResource.ID, nil
```
Probably don't need `resourceGroupID` var

File: pkg/actuators/client/client.go
Comment (srcarrier) [inline pkg/actuators/client/client.go:381]:
```suggestion
				return *eachSubnet.ID, nil
```

File: pkg/actuators/client/client.go
Comment (elmiko) [reply pkg/actuators/client/client.go:141]: are we talking about the kubernetes client or the ibmcloud client? (just want to make sure i'm following)

File: pkg/actuators/client/client.go
Comment (gsrajpal) [reply pkg/actuators/client/client.go:312]: Added that just to keep it aligned with VPC custom image creation. Any custom image we create is `private`. Not quite sure what the default visibility is when we pull a `List of images` through SDK, but if it's set to `public`, we will not get the image. So setting it should be better so we know we are always pulling `private` image. 

If VPC ever allows making that change while creating a custom image, we will still be good unless the default becomes `public` during that creation and we are not setting the `visibility` - at that point will need to update Terraform scripts as well.

File: pkg/actuators/client/client.go
Comment (gsrajpal) [reply pkg/actuators/client/client.go:342]: Agree, we can `SetName` here. Didn't realize.

File: pkg/actuators/client/client.go
Comment (srcarrier) [reply pkg/actuators/client/client.go:165]: Up to you Gagan - just pointing out that it shouldn't be necessary.

File: pkg/actuators/client/client.go
Comment (srcarrier) [reply pkg/actuators/client/client.go:141]: The `IbmClientBuilder` client within the machine actuator that's instantiated in the main function of /cmd/manager/main.go. The service endpoint (based on region) should be the same for a given OCP cluster.

File: pkg/actuators/client/client.go
Comment (srcarrier) [inline pkg/actuators/client/client.go:420]: Since the `SecurityGroupIdentityList` slice references the make array of size `len(securityGroupNames)`, won't these lengths be equal even if you only found 1 of 2 subnet IDs?
Also, `retrived` is misspelled in the above comment.

File: pkg/actuators/client/client.go
Comment (elmiko) [reply pkg/actuators/client/client.go:141]: if the service endpoint won't change during operation, then i think it would be safe to keep the same client around.

Comment (gsrajpal) [reply pkg/actuators/client/client.go:141]: Sounds good. Will move it to the client.

Question -- Do we need to account for multi `region` Machine provisioning? From the sounds of it, I believe we should not be concerned about `regions` here.

- I remember you mentioned that machines in different regions are not grouped at the MachineSet level, but we can have them.

...Probably my last doubt around this. Will make changes accordingly.

File: pkg/actuators/machine/conditions.go
Comment (srcarrier) [inline pkg/actuators/machine/conditions.go:62]: The `Status` delta check here between current vs new to set the transition time should occur prior to the assignment of the current status to new on line 56

File: pkg/actuators/machine/reconciler.go
Comment (srcarrier) [inline pkg/actuators/machine/reconciler.go:113]: Might be good to log an error here if err != nil

File: pkg/actuators/client/client.go
Comment (elmiko) [reply pkg/actuators/client/client.go:141]: as i understand it, it is acceptable to have a cluster with machines in several regions. those regions will be gated at the machineset level.

so, if you need a different client to interact with the different regional endpoints, then you would have to create the client for each machine. i would think.

File: pkg/actuators/machine/reconciler.go
Comment (gsrajpal) [reply pkg/actuators/machine/reconciler.go:113]: The controller should log if `err != nil`

Catching and returning the err in `InstanceExistsByName()` in `client.go`

..If we log here then we will be logging it twice.

File: pkg/actuators/client/client.go
Comment (gsrajpal) [reply pkg/actuators/client/client.go:420]: Nice catch - Updating the Slice len() to 0

File: pkg/actuators/client/client.go
Comment (srcarrier) [reply pkg/actuators/client/client.go]: In the aws provider, clients are instantiated within the [machine scope](https://github.com/openshift/cluster-api-provider-aws/blob/c69a3acaee27ccbaf163d5955ac727c10f2f49e0/pkg/actuators/machine/machine_scope.go#L68) with the region being fetched from the provider spec therein. That approach should ensure that a client is leveraging the proper endpoint based on the region within the machine scope.

**Review by srcarrier — CHANGES_REQUESTED** [review]
Code issues (e.g. [example](https://github.com/openshift/cluster-api-provider-ibmcloud/pull/3#discussion_r661444975)) should be addressed before merging.
It'd be good to sort out the service endpoint issue as well since the defaulting to `us-south` won't blow up the code ~ it may prove more difficult to determine the cause of the behavior problems resulting from vpc service calls to the wrong region (at least for folks not using `us-south`).
Regarding code style suggestions/comments, I leave that up to you. They're just suggestions.

**Comment by gsrajpal** [comment]
Thank you for taking a look @srcarrier 

I pushed the changes that we talked about above, have a look: [New Commit](https://github.com/openshift/cluster-api-provider-ibmcloud/pull/3/commits/4095f9cd0561d5c5697c7ec89552a9ceebc53ef7) 

...will make further changes if needed.

File: pkg/actuators/client/client.go
Comment (srcarrier) [inline pkg/actuators/client/client.go:394]: You could have an `else` statement here, which would inform us as to which specific security group name was not found in the list. We can still do the `len` check below, but the resulting err msg doesn't tell us which specific names were not found. Could also just return an error here.
Could also have an `else` for the `if securityGroups != nil` and error there as well, indicating that no security groups were retrieved from vpc.

**Review by srcarrier — APPROVED** [review]
Made a small suggestion around the retrieval of subnets _(see last comment)_ and how we're logging subnets that aren't found.
It'd be good if the logs told us specifically which subnets weren't found instead of giving us the full list, saving us from manually having to make that determination.
Aside from that, this PR looks good 👍

File: pkg/actuators/client/client.go
Comment (gsrajpal) [reply pkg/actuators/client/client.go:396]: I just updated the logic to account for missing security group ID's 

We should get a clear log message indicating the missing ID now:

[Commit](https://github.com/openshift/cluster-api-provider-ibmcloud/pull/3/commits/85b9729a376592c4f7cba33279be6a06e319514e)

**Pull request merged**
